### DISCUSSION TITLE
[API Proposal]: Why can't access a constructor element using 'this'

### DISCUSSION THREAD

**Discussion by Mr0N** [dotnet/csharplang#7715] [General] [answered]
### Background and motivation

https://sharplab.io/#gist:5ec979cc565a8a49ed04f005f260c7e1
Why can't access a constructor element using 'this' ?

### API Proposal

```csharp
  class One(string one)
  {

      public void Start()
      {
          Console.WriteLine(one);
          //Console.WriteLine(this.one); //error

      }
  }
```

### API Usage

-

### Alternative Designs

-

### Risks

-

**Comment by GalaxiaGuy** [comment, marked as answer]
Because they are parameters, not instance members:

https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/primary-constructors#primary-constructors

> It's important to view primary constructor parameters as parameters even though they are in scope throughout the class definition.

**Reply by Mr0N** [reply]
well, it's somehow strange that the fields are available in the entire instance. How to distinguish these parameters from other members of the instance?

**Reply by CyrusNajmabadi** [reply]
They are not fields.  They are parameters.  Their scoping makes them available across teh entire type (just like a method parameter is available across the entire method).  

> How to distinguish these parameters from other members of the instance?

There's no way.  Like there's no way to distinguish a method parameter from other members (i.e. locals) within the method.

**Reply by Mr0N** [reply]
ok, I understand, don’t need "this", but I don’t like it all. The very logic is that parameters and fields are different, taking into account that there is a field under the hood. But why these parameters, it’s not easier to make a private field simply, everything will be like in js, but then it just confuses everything.

Thai  somehow it is not good, if even parameters, then why not make some keyword to access them, for example "this", only for parameters. This will not violate the old code that was written before.

**Reply by Mr0N** [reply]
Well, the main problem is compiled assemblies, how to understand what fields are there. And how to understand what I am now referring to in the class, to a parameter or, for example, to a public field in the class that was not written by me?

**Reply by CyrusNajmabadi** [reply]
> , taking into account that there is a field under the hood.

This is an implementation detail, outside of hte language.  Compilers are free to implement this however they want, including not using fields.

> But why these parameters

Because in many cases you just need the parameter, and there is no need to designate explicit storage for it (or do things like explicit assignment).  For very common cases this is much simpler. 

>  then why not make some keyword to access them

Because there isn't a need to.  THey can already be accessed by name.  Just like all other parameters.

**Reply by CyrusNajmabadi** [reply]
> Well, the main problem is compiled assemblies, how to understand what fields are there.

These are private impl details.  They aren't included, and you wouldn't know about them (for example, in a ref-assembly).  THere  is also no need to know as they can change at any time.

> And how to understand what I am now referring to in the class, 

This is the same as it has always been in c#.  When you are in a local function, you can't tell if you're referencing a parameter or a local.  You have to just go and see.  

Many things in the language have no way to reference themselves except by simple name.  You're free to suggest a new referencing mechanism.  But we haven't seen a need for that so far.

**Reply by Mr0N** [reply]
This is not true, with the help of the this keyword you can clearly say that the reference is not to the internal changes of the method.

**Reply by Mr0N** [reply]
Now it becomes very confusing code that is difficult to read. It is not clear what this is, this is a parameter, this is a public field from a class that this class inherits from, this is a change in a method.

**Reply by CyrusNajmabadi** [reply]
> This is not true, 

It is true.  you cannot use `this` to distinguish between locals and parameters within a method or local function.  Neither are members on the instance value.

**Reply by CyrusNajmabadi** [reply]
> Now it becomes very confusing code that is difficult to read.

It's always been the case that distinguishing between parameters and locals was not possible at hte use site.

Again, if you want a feature to support that, feel free to suggest it.  But right now, as these are just parametres, they behave hte same way parameters always have wrt to lookup, and they have that same limitation that has existed since c# 1.0.

**Reply by HaloFour** [reply]
Captures for lambdas and local functions are also typically implemented as fields, and you could never qualify them as such.  That's the case even though the lambda body and captures are emitted in the same separate type.  Just because the implementation detail is to emit a field doesn't mean that the language should treat them as such.

If you don't like it, you don't have to use the feature.

**Comment by Mr0N** [comment]
> Because they are parameters, not instance members:
> 
> https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/primary-constructors#primary-constructors
> 
> > It's important to view primary constructor parameters as parameters even though they are in scope throughout the class definition.

![image](https://github.com/dotnet/runtime/assets/61237140/eb7651cd-a8b0-4762-8129-3df58898d75f)

**Reply by CyrusNajmabadi** [reply]
That is the compiled code, which is an implementation detail.  THe language does not care how it is implemented, and the compiler is free to implement it in other ways (including not using a field).

**Comment by Mr0N** [comment]
And I understood that the studio somehow magically replaces these symbols <> with others. Why is it done this way?

**Reply by CyrusNajmabadi** [reply]
Because they are implementation details, not to be used/referenced by other code.  We give 'mangled' names like these for all the things we generate as impl details.

**Comment by Mr0N** [comment]
![image](https://github.com/dotnet/runtime/assets/61237140/5005ef28-144b-4978-b9da-dcf5b8502179)
and what does this mean, and if the class is large, and I add this field, the whole program will eventually break

**Reply by CyrusNajmabadi** [reply]
It means you have a parameter *and* a field with the same name.  Which is something that has been legal/normal since c# 1.0.  Indeed, the common way to write types is to have a field with a particular name, then a parameter constructor with the same name, then assign the parameter constructor to that field.  i.e.:

```c#
class C
{
    int x;
    public C(int x)
        => this.x = x;
}
```

That can now be written with primary constructors as:

```c#
class C(int x)
{
    int x = x;
}
```

**Reply by Mr0N** [reply]
It's cool, but there is a problem here, the behavior of the class changes relative to the change of the name of the parameter in the constructor. If, for example, there is a base class with this field, what should be done?

**Reply by Mr0N** [reply]
![image](https://github.com/dotnet/csharplang/assets/61237140/312e8f4d-0c8c-40ea-842a-8c05ed6b31e0)
https://sharplab.io/#gist:3ba3b877c03fa0b7eac21acdd32893b3
What to do with this now? Suppose that the class is in some dll assembly

**Reply by HaloFour** [reply]
You're not using the primary constructor parameter, therefore it is not captured.  That has nothing to do with whether or not you have a derived class.

**Reply by Mr0N** [reply]
![image](https://github.com/dotnet/csharplang/assets/61237140/c100f736-ddbb-4aa3-9fd3-3cb72f492240)

**Reply by HaloFour** [reply]
I see what you're referring to.  Yes, the compiler will prefer the field defined in the parent class over making a capture of the primary constructor parameter.  This is a known behavior and discussed during design meetings.

**Reply by HaloFour** [reply]
Here is the part of the spec that discusses this:  https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/primary-constructors.md#warn-on-shadowing-by-a-member-from-base

As you can see it does result in a warning.

**Reply by Mr0N** [reply]
why not an error?

**Reply by Mr0N** [reply]
How then to get information from the constructor, if there is a similar field

**Reply by HaloFour** [reply]
> why not an error?

Because it's not illegal.

> How then to get information from the constructor, if there is a similar field

Assign it explicitly to a field in an initializer.

**Comment by GalaxiaGuy** [comment]
The existence of a field is an implementation detail that the C# compiler can choose. The names are chosen to intentionally not collide with something a user can enter.

https://devblogs.microsoft.com/dotnet/check-out-csharp-12-preview/

> For primary constructor parameters, the capture is implemented by generating a private backing field on the class or struct itself. The field has an “unspeakable” name, which means it will not collide with other naming and is not obvious via reflection.

Many C# features are implemented this way.

If you add a field with a same name, one will shadow the other. References to `one` in your example will refer to the field and not the parameter.

https://dotnetfiddle.net/QolwsW

**Comment by Mr0N** [comment]
![image](https://github.com/dotnet/runtime/assets/61237140/fe80839d-e4b3-4ac2-ab66-c87681692489)
this is where the problem turns out

**Reply by CyrusNajmabadi** [reply]
yes. the same happens with normal fields as well.

**Comment by Mr0N** [comment]
> The existence of a field is an implementation detail that the C# compiler can choose. The names are chosen to intentionally not collide with something a user can enter.
> 
> https://devblogs.microsoft.com/dotnet/check-out-csharp-12-preview/
> 
> > For primary constructor parameters, the capture is implemented by generating a private backing field on the class or struct itself. The field has an “unspeakable” name, which means it will not collide with other naming and is not obvious via reflection.
> 
> Many C# features are implemented this way.
> 
> If you add a field with a same name, one will shadow the other. References to `one` in your example will refer to the field and not the parameter.
> 
> https://dotnetfiddle.net/QolwsW

It's easier to simply restrict the user from entering a field identical to the primer in the constructor; it will help avoid potential issues

**Comment by Mr0N** [comment]
> ![image](https://user-images.githubusercontent.com/61237140/285234729-fe80839d-e4b3-4ac2-ab66-c87681692489.png) this is where the problem turns out

![image](https://github.com/dotnet/runtime/assets/61237140/db707074-9efa-4e2b-8eef-524748d9f6df)
does

**Comment by Mr0N** [comment]
Such code becomes very difficult to read
```
using System;
using System.Web;

(new One(21) as Two).Start();
  class One(int one):Two(one)
 {
    public override void Start()
    {
        Console.WriteLine(one);
     //   base.Start();
    }
    private int one =7382;
 }
 class Two(int one)
 {
    public virtual void Start(){
    
        Console.WriteLine(one);
    }
 }
```

**Comment by Mr0N** [comment]
https://sharplab.io/#gist:a276efd01fb71475706fed5a53d63462
Why does the compiler change the variable name here?
![image](https://github.com/dotnet/runtime/assets/61237140/ebd5f053-eb06-444e-a54d-0bd4e61a8c56)

**Reply by CyrusNajmabadi** [reply]
locals dont' have names in IL.  What you're looking at is the arbitrary name the decompiler picked for you.

**Comment by HaloFour** [comment]
Primary constructor parameters are not members/fields, nor can they be accessed as members/fields.  If they are used beyond the lifecycle of the constructor then they are captured, but how they are captured is an implementation detail.  The fact that the implementation detail might use fields as storage isn't relevant to the language.

This behavior is intentional.

**Reply by Mr0N** [reply]
That is, according to the idea, it is not possible to get access to the fields of the constructor primer?

**Reply by HaloFour** [reply]
You access the captured parameters by their name.  They are not members, so you can't reference them as members.  There are no fields as far as the language is concerned.

**Reply by Mr0N** [reply]
At first glance, the fields seem like those in JavaScript, but without using "this."

**Reply by CyrusNajmabadi** [reply]
> At first glance, the fields seem

These are not fields.  

> but without using "this."

Correct.  They are not members.  They are are parameters.  They are not accessed off of 'this'

**Comment by Mr0N** [comment]
https://sharplab.io/#gist:1233568c9eea11dc36412dfdcb090d3e
![image](https://github.com/dotnet/csharplang/assets/61237140/f9e07dd1-21c0-4bf5-9398-63d75928c4ca)
How to understand where the information is being recorded here?

**Reply by HaloFour** [reply]
By reading the code?  I don't know what your intent is from this rapid-fire picture show approach to trying to discuss the feature.  The behaviors have been discussed during design reviews and publicly on this repository before the feature was shipped.

If you want to suggest new ways of prefixing/qualifying identifiers then do so.  The warning exists to try to inform you of the fact that you have a parameter that shadows an accessible field.

**Reply by Mr0N** [reply]
How to read it, Base is a class of some kind of compiled assembly of downloaded from nuget

**Reply by HaloFour** [reply]
The compiler will warn about the shadowing regardless of where the base class is declared.

Beyond that, I don't know what you're asking.  The questions around the behavior of the feature have been answered.  If you don't like them, then you're welcome to open discussions around ways of changing them.  However, as the feature has shipped, short of something drastic and unexpected it's unlikely that anything will change.
